pcsx2/GS/Renderers/HW/GSTextureCache.cpp

```diff
@@ -2213,21 +2206,55 @@ void GSTextureCache::Target::Update()
 
 	// m_renderer->m_perfmon.Put(GSPerfMon::Unswizzle, w * h * 4);
 
-	// Copy the new GS memory content into the destination texture.
-	if (m_type == RenderTarget)
+	// The idea: If we have more than one dirty rectangle, we upload the whole thing, and then
+	// copy/upscale the specific areas we want into it.
+	if (m_dirty.size() > 1)
 	{
-		GL_INS("ERROR: Update RenderTarget 0x%x bw:%d (%d,%d => %d,%d)", m_TEX0.TBP0, m_TEX0.TBW, r.x, r.y, r.z, r.w);
+		const GSVector2i bs = GSLocalMemory::m_psm[m_TEX0.PSM].bs;
+		for (const auto& dr : m_dirty)
+		{
+			// dirty area in unscaled space
+			GSVector4i drc(dr.GetDirtyRect(m_TEX0));
```

Comment:
Just to clarify a point, the `GSDirtyRect::GetDirtyRect` method used here (and it was the same before these changes) is a shortcut that properly works only in some cases.

Given the lack of pitch conversion you need to have buffer width compatibility here (between the draw and the target), also this code supposes that the writes occur with compatible memory mapping format (#5534 tries [and fails] to surpass this).
There's an issue with the pitch (`m_TEX0.TBW`) attached to our framebuffers: the original one is overwritten with this of the last write that is reported in the m_dirty of the same target (possibly trashing the old value) without any pitch conversion occurring in the process.
I fear that by being more precise here with the uploads, these limitations will result in some issues.

Also there's a chance that another draw "overwrites" the local data but it is not reported to the actual target nor its dirty rectangles list, hence even with this upgraded logic, some issues could persist due to the `m_dirty` referencing invalid data (`GSTextureCache::InvalidateVideoMem` not really invalidating). The only sane way to solve this properly seems to keep a data structure tracking each page/(block?) state (like I tried to do in my branch, if you wanna take a look https://github.com/iMineLink/pcsx2/blob/gs/hw-tc-page-ownership/pcsx2/GS/Renderers/HW/GSTextureCache.h#L159).

In conclusion, this is a step forward towards compatibility in my opinion as the upload is restricted to the dirty areas and not to a generic union, but it could make more obvious other limitations of the current caching system.

Consider making the optimizations done in the `GSTextureCache::Source::Write` to merge togheter the dirty rectangles and reduce the number of uploads.